py/desitarget/brightstar.py

```diff
@@ -453,22 +660,208 @@ def is_in_bright_star(targs,starmask):
+    radius = 1.001*starmask["IN_RADIUS"]/60.
+
+    #ADM determine the area of each mask
+    area = cap_area(radius)
+
+    #ADM determine the number of SAFE locations to assign to each
+    #ADM mask given the passed number of locations per sq. deg.
+    Nsafe = np.ceil(area*Npersqdeg).astype('i')
+
+    #ADM determine Nsafe Dec offsets equally spaced around the perimeter for each mask
+    offdec = [ rad*np.sin(np.arange(ns)*2*np.pi/ns) for ns, rad in zip(Nsafe,radius) ]
+
+    #ADM use offsets to determine DEC positions
+    dec = starmask["DEC"] + offdec
+
+    #ADM determine the offsets in RA at these Decs given the mask center Dec
+    offrapos =  [ sphere_circle_ra_off(th,cen,declocs) for th,cen,declocs in zip(radius,starmask["DEC"],dec) ]
+
+    #ADM determine which of the RA offsets are in the positive direction
+    sign = [ np.sign(np.cos(np.arange(ns)*2*np.pi/ns)) for ns in Nsafe ]
+
+    #ADM determine the RA offsets with the appropriate sign and add them to the RA of each mask
+    offra = [ o*s for o,s in zip(offrapos,sign) ]
+    ra = starmask["RA"] + offra
+
+    #ADM have to turn the generated locations into 1-D arrays before returning them
+    return np.hstack(ra), np.hstack(dec)
+
+
+def append_safe_targets(targs,starmask,drstring=None,drbricks=None):
+    """Append targets at SAFE (BADSKY) locations to target list, set bits in TARGETID and DESI_TARGET
+
+    Parameters
+    ----------
+    targs : :class:`recarray`
+        A recarray of targets as made by desitarget.cuts.select_targets
+    starmask : :class:`recarray`
+        A recarray containing a bright star mask as made by desitarget.brightstar.make_bright_star_mask
+    drstring : :class:`str`, optional
+        The imaging data release for the targets as a 9-string (i.e. 'dr3      ') to update TARGETID
+        If this is not passed, then SAFE (BADSKY) targets will be generated, but target bits will not be meaningful
+    drbricks : :class:`recarray`, optional
+        A rec array containing at least the "ra", "dec" and "nobjs" columns from a survey bricks file for the passed 
+        Data Release. This is typically used for testing only, as the code can determine the NERSC location of 
+        a survey bricks file directly from drstring
+
+    Returns
+    -------
+        The original recarray of targets (targs) is returned with additional SAFE (BADSKY) targets appended to it
+
+    Notes
+    -----
+        - See the Tech Note at https://desi.lbl.gov/DocDB/cgi-bin/private/ShowDocument?docid=2346 for more details
+          on the SAFE (BADSKY) locations
+        - See the Tech Note at https://desi.lbl.gov/DocDB/cgi-bin/private/RetrieveFile?docid=2348 for more details
+          on setting the SKY bit in TARGETID
+        - Currently hard-coded to create an additional 10,000 safe locations per sq. deg. of mask. What is the 
+          correct number per sq. deg. (Npersqdeg) for DESI is an open question.
```

Comment:
Can you explain the concept behind Npersqdeg more?  I thought every mask would get a set of safe locations at its perimeter and that if a density was specified it would be a linear density (i.e. how far apart on the perimeter can they be?) rather than an area density.